sensu/sensu-go
# Metric Threshold Service Checks

* Define metric threshold schema
  * MetricThreshold
  * MetricThresholdRule
  * MetricThresholdTag
* Evaluate metric threshold rules for checks with output metrics

Signed-off-by: Francis Guimond <fguimond@users.noreply.github.com>

Closes #4533 
Closes #4534 

## What is this change?

Extend the capabilities of Sensu service check execution and output metric processing engine (in the Sensu Agent) to add built-in support for analyzing check output metrics and overriding the check status based on one or more metric thresholds.

* Create a schema to add the metric thresholds to the CheckConfig
* Evaluate the metric thresholds when necessary

## Why is this change necessary?

Native “output metric threshold” tooling would immediately turn any source of gauge-style1 metrics into a viable service check (e.g. Prometheus exporters), enabling customers to get real-time alerts for metrics collected by Sensu.

## Does your change need a Changelog entry?

Yes.

## Do you need clarification on anything?

No.

## Were there any complications while making this change?

Generating the schema using `protoc` was problematic since my project is not in `~/go/src/github.com/...`. I had to temporarily copy the code back and forth to generate the schema. A fix probably would be a good idea.

## Have you reviewed and updated the documentation for this change? Is new documentation required?

@hillaryfraley - new documentation will be required.

## How did you verify this change?
### Unit tests
### Manual testing
* Setup `node_exporter` on your workstation
* Add the `unix-metrics` subscriptions to one of your entity
* Create the following check
```
---
type: CheckConfig
api_version: core/v2
metadata:
  name: node_exporter
  namespace: default
spec:
  command: curl http://127.0.0.1:9100/metrics
  interval: 15
  publish: true
  subscriptions:
    - unix-metrics
  output_metric_format: prometheus_text
  output_metric_thresholds:
  - name: node_load1
    tags: null
    thresholds:
    - max: "2.0"
      min: ""
      null_status: 0
      status: 1
    - max: "3.0"
      min: ""
      null_status: 0
      status: 2
  - name: node_filesystem_avail_bytes
    tags:
    - name: device
      value: /dev/disk1s1s1
    - name: fstype
      value: apfs
    thresholds:
    - max: ""
      min: "500e+9"
      null_status: 0
      status: 1
    - max: ""
      min: "250e+9"
      null_status: 2
      status: 2
```
* Observe the check results in the web ui
* Try new scenarios by editing the check using the following command: `sensuctl edit core/v2.CheckConfig node_exporter`

## Is this change a patch?

No.
